### PR TITLE
Kani regression

### PR DESCRIPTION
**Pull request by markrtuttle**
This pull request adds regression tests that run kani --visualize to test viewer on kani tests.

This pull request also does some minor cleanup, resets required python version back to 3.6 and advances version number to 3.2.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: tests/kani/Makefile [hunk 1]
```diff
@@ -0,0 +1,38 @@
+SHELL=/bin/bash
+
+# Assume cbmc and cbmc-viewer are installed in PATH
+#   skip ./kani/scripts/setup/ubuntu/install_cbmc.sh
+#   skip ./kani/scripts/setup/install_viewer.sh 2.11
+
+KANI=kani
+
+# ARCH is ubuntu or macos-10.15
+ARCH=ubuntu
```
Comment (karkhaz) [inline tests/kani/Makefile:10]: Is this expected to be modified by the user? If so, shouldn't we use `?=` so that the environment variable overrides it if set?

File: tests/kani/Makefile [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+clone:
+	git clone https://github.com/model-checking/kani.git $(KANI)
+	cd $(KANI) && git submodule update --init
+
+# MacOS warning: If brew fails trying to install ctags over
+# universal-ctags, then replace ctags with universal-ctags
+# ./kani/scripts/setup/macos-10.15/install_deps.sh
+
+build:
+	PATH=$$HOME/.cargo/bin:$$PATH && cd $(KANI) && \
```
Comment (karkhaz) [inline tests/kani/Makefile:21]: This style makes it difficult to see which of these commands failed. Suggest:
```
build:
	PATH=$$HOME/.cargo/bin:$$PATH  $(MAKE) -C $(KANI) _build

_build:
	./scripts/setup/$(ARCH)/install_deps.sh
	./scripts/setup/install_rustup.sh
	cargo build --workspace
```

File: tests/kani/Makefile [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+	    ./scripts/setup/$(ARCH)/install_deps.sh && \
+	    ./scripts/setup/install_rustup.sh && \
+	    cargo build --workspace
+
+test:
+	PATH=$$HOME/.cargo/bin:$$(pwd)/$(KANI)/scripts:$$PATH && \
+	echo cbmc --version: $$(cbmc --version) && \
```
Comment (karkhaz) [inline tests/kani/Makefile:28]: Make prints the commands out before running them, so this is going to print out
```
echo cbmc --version: $$(cbmc --version)
cbmc --version: 5.54
```
I suggest either `@echo` so that the commandline itself is not printed out. Or, simply write
```
cbmc --version
```
which would then print out
```
cbmc --version
5.54
```

Also, for diagnostic output, it's better to print to stderr, like
```
>&2 echo foo
```

File: tests/kani/Makefile [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+	echo cbmc-viewer --version: $$(cbmc-viewer --version) && \
+	echo kani --version: $$(kani --version) && \
+	./write_build_ninja.py $(KANI)/tests/kani && \
+	ninja
+
+clean:
+	$(RM) -r *~ build.ninja .ninja_log __pycache__
```
Comment (karkhaz) [inline tests/kani/Makefile:35]: What is `*~` for? Some text editors (emacs) save swap files with this extension, this seems to be trashing those which is unrelated to cleaning out the build? Or does the build itself dump out files like this?

File: tests/kani/write_build_ninja.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
+#!/usr/bin/env python3
```
Comment (karkhaz) [inline tests/kani/write_build_ninja.py:1]: These files all need copyright headers

File: tests/kani/write_build_ninja.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    tags = ['ignore', 'fixme', 'tests/kani/ForeignItems/main.rs']
+    return [src for src in sources if all(tag not in src for tag in tags)]
+
+def kani_pragma(pragma, rust_code):
+    # Examples of kani pragmas:
+    # // kani-verify-fail
+    # // kani-check-fail
+    # // kani-flags: --default-unwind 8
+    try:
+        regexp = fr'//[ \t]*{pragma}([^\n]*)'
```
Comment (karkhaz) [inline tests/kani/write_build_ninja.py:33]: `\s*` rather than `[ \t]*`?

File: tests/kani/write_build_ninja.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    source = Path(source)
+    with open(source, encoding='utf-8') as src:
+        rust_code = src.read()
+    if kani_verify_fail(rust_code) or kani_check_fail(rust_code):
+        return None
+
+    # Warning: concurrent invocations of kani in a single directory
+    # should generate race conditions that have so far been
+    # unobserved.  The count variable here is just a hack just to get
+    # ninja to run.
```
Comment (karkhaz) [inline tests/kani/write_build_ninja.py:57]: So use a pool?

File: tests/kani/write_build_ninja.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+            'inputs':  str(source),
+            'variables': {'dir': str(source.parent),
+                          'src': str(source.name),
+                          'flags': kani_flags(rust_code)}}
+
+def build_ninja_file(sources, build='build.ninja'):
+    with open(build, 'w', encoding='utf-8') as output:
+        writer = ninja_syntax.Writer(output)
+        writer.rule(RULE_NAME, 'cd "${dir}" && kani ${flags} --visualize ${src}')
+        for source in sources:
```
Comment (karkhaz) [inline tests/kani/write_build_ninja.py:71]: Is it really true that there's just one kani test per source file?

Here's how i would do this. For each source file, if it contains `check-fail` or `verify-fail`, return an empty list rather than none. Else return a list of the dicts that you're currently returning. Then the code here becomes:
```
for source in sources:
  for build in ninja_build_statements(source):
      writer.build(**build)
```
No need for this try-catch, which also masks other potential errors. It also allows for multiple sets of kani invocations per source file, if that's something possible/desirable.

File: tests/kani/Makefile (hunk 1 above)
Comment (markrtuttle) [reply tests/kani/Makefile:10]: Done.  I have been setting it from the make command.

File: tests/kani/Makefile (hunk 2 above)
Comment (markrtuttle) [reply tests/kani/Makefile:21]: Done something similar.  The `-C $(KANI)` won't work because it requires a second Makefile in the `$(KANI)` directory.

File: tests/kani/Makefile (hunk 3 above)
Comment (markrtuttle) [reply tests/kani/Makefile:28]: Done something similar.

File: tests/kani/Makefile (hunk 4 above)
Comment (markrtuttle) [reply tests/kani/Makefile:35]: The editor emacs makes backup files between saves with names having `~` appended.  I'm going to leave this because it is common cleanup work at the end of hacking.

File: tests/kani/write_build_ninja.py (hunk 6 above)
Comment (markrtuttle) [reply tests/kani/write_build_ninja.py:33]: No, `\s` will match against a newline `\n` and I want to match to end of the "line".

File: tests/kani/write_build_ninja.py (hunk 8 above)
Comment (markrtuttle) [reply tests/kani/write_build_ninja.py:71]: I've implemented this suggestion.

File: tests/kani/write_build_ninja.py (hunk 5 above)
Comment (markrtuttle) [reply tests/kani/write_build_ninja.py:1]: Done.

File: tests/kani/write_build_ninja.py (hunk 7 above)
Comment (markrtuttle) [reply tests/kani/write_build_ninja.py:57]: That would require one pool per directory.  I'm going to leave this now and add this as a TODO in the code.

Comment (markrtuttle) [reply tests/kani/write_build_ninja.py]: I implemented the pools, it wasn't that hard.
